feat: added new inactivity wrapper component

al-rosenthal opened the pull request:
# Summary of Changes

Created a new component: `InactivityWrapper.tsx`. This component is intended to wrap the app view to track user interactions (clicks, swipes, ect.). When an interaction is detected a timer internal to the component resets, if no interactions in a set time are made while the app is open, the user is logged out and is sent back to the pin entry screen. 

# Related Issues

[bc-wallet-mobile: 2136](https://github.com/bcgov/bc-wallet-mobile/issues/2136)

# Pull Request Checklist

Tick all boxes below to demonstrate that you have completed the respective task. If the item does not apply to your this PR **check it anyway** to make it apparent that there's nothing to do.

- [x] All commits contain a DCO `Signed-off-by` line (we use the [DCO GitHub app](https://github.com/apps/dco) to enforce this);
- [x] Updated LICENSE-3RD-PARTY.md for any added dependencies or vendored components;
- [x] Updated documentation as needed for changed code and new or modified features;
- [x] Added sufficient [tests](../__tests__/) so that overall code coverage is not reduced.

If you have _any_ questions to _any_ of the points above, just **submit and ask**! This checklist is here to _help_ you, not to deter you from contributing!

Pro Tip 🤓

- Read our [contribution guide](../CONTRIBUTING.md) at least once; it will save you a few review cycles!
- Your PR will likely not be reviewed until all the above boxes are checked and all automated tests have passed.

_PR template adapted from the Python attrs project._

al-rosenthal committed:
feat: added new inactivity wrapper component

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

jleach commented on packages/legacy/core/App/components/misc/InactivityWrapper.tsx:
```diff
@@ -0,0 +1,121 @@
+
+const InactivityWrapper: React.FC<PropsWithChildren<InactivityWrapperProps>> = ({ children, timeoutLength }) => {
+  const [logger] = useServices([TOKENS.UTIL_LOGGER])
+  useStore()
```
Any reason to call `useStore()` here?

al-rosenthal replied:
nope

jleach commented on packages/legacy/core/App/components/misc/InactivityWrapper.tsx:
```diff
@@ -0,0 +1,121 @@
+import { DispatchAction } from '../../contexts/reducers/store'
+import { TOKENS, useServices } from '../../container-api'
+
+export enum LockOutTime {
```
Enums are somewhat of a hack in typescript. I think we favour `as const` like

```
const X = {
  Blah: 'zzzzzz',
  Blarb: 'rrrrr',
} as const
```

al-rosenthal replied:
Happy to change that. Is there a way to still have this object act as a type when defining something like a type or interface? for example:

```typescript
const MyConst = {
  Blah: 'something'
} as const

type MyType = {
  someAttribute: MyConst
}

```

jleach commented on packages/legacy/core/App/components/misc/InactivityWrapper.tsx:
```diff
@@ -0,0 +1,121 @@
+  Never = 0,
+}
+
+interface InactivityWrapperProps {
```
Type or Interface? I tend to favour `type` but maybe interface is better?

Re: https://www.youtube.com/watch?v=zM9UPcIyyhQ

al-rosenthal replied:
I have no strong feelings towards either, it is a type for now.

jleach commented on packages/legacy/core/App/components/misc/InactivityWrapper.tsx:
```diff
@@ -0,0 +1,121 @@
+  const [_, dispatch] = useStore()
+  const { agent } = useAgent()
+  const { removeSavedWalletSecret } = useAuth()
+  const timeout_minutes = timeoutLength !== undefined ? timeoutLength : LockOutTime.OneMinute
```
Favour cammelCase over snake_case as a convention and descriptive variable names like `timeoutInMinues`

al-rosenthal replied:
updated

jleach commented on packages/legacy/core/App/components/misc/InactivityWrapper.tsx:
```diff
@@ -0,0 +1,121 @@
+    }
+  }
+
+  const minutesToMilliseconds = (minutes: number) => {
```
I might just simplify this to `const blahAsMiliseconds = minutes * 60000` at the top of the file.

al-rosenthal replied:
good call

jleach commented on packages/legacy/core/App/components/misc/InactivityWrapper.tsx:
```diff
@@ -0,0 +1,121 @@
+    })
+  ).current
+
+  const resetInactivityTimeout = (minutes: number) => {
```
You probably need to wrap all your fn inside a `useCallback`. Every time a component renders a function will be re-created which is a performance hit. Also, because a fn can be a dependency on a `useEffect` when it is re-created it will change causing a `useEffect` to trigger. So, in general, its good to just wrap 'em and let RN sort it out.

al-rosenthal replied:
functions are wrapped in callbacks

al-rosenthal committed:
refactor: resolving PR comments

modified variable names, added wrapper to demo app

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
refactor: wrapped functions in useCallback hooks

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
refactor: interface -> type

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
refactor: code quality issues

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
Merge branch 'main' into feat/AutoLockComponent

al-rosenthal committed:
refactor: refactored wrapper

wrapped functions in calbacks and added lockout logic when app is in background

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
refactor: refactored root stack

removed old lockout code from root stack

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
chore: linter fixes

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

bryce-mcmath reviewed (commented):
This looks great but you may want to check with Kim / Clecio if an actual settings option (where the user can switch between the different lockout times) was part of the ticket.

al-rosenthal committed:
feat: added new lockout screen

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
feat: wired up lockout timer to settings screen

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
fix: fixing lint issues

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
feat: updated store to handle auto lock

renamed auto lock screen, updated enum name, updated store types

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
fix: updated string to accoutn for dynamic lock period

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
fix: fixed timer loading from stroage

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
fix: fixed lint and typecheck errors

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
Merge branch 'main' into feat/AutoLockComponent

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
docs: added test labels

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

bryce-mcmath commented on packages/legacy/core/App/contexts/reducers/store.ts:
```diff
@@ -646,6 +665,7 @@ export const reducer = <S extends State>(state: S, action: ReducerAction<Dispatc
       }
     }
     default:
+      console.log(`${action.type} is not available`)
```
Can be helpful for local dev but shouldn't be committed, the linter will complain about `console` usage

al-rosenthal replied:
removed

bryce-mcmath commented on packages/legacy/core/App/localization/en/index.ts:
```diff
@@ -248,7 +248,7 @@ const translation = {
     "BiometricsError": "Biometrics were not successful.",
     "BiometricsErrorEnterPIN": "Please enter your wallet PIN.",
-    "LockedOut": "You locked out after 5 minutes of inactivity.",
+    "LockedOut": "You were locked out after ## minute(s) of inactivity.",
```
There's an example of how to do this without using `String.prototype.replace` if you check the "MinutesAgo" entry.

al-rosenthal replied:
swapped out with established pattern

bryce-mcmath reviewed (changes requested):
Looks really good, just two tiny changes and a rebase then I will approve

al-rosenthal committed:
chore: added translations for new text

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
Merge branch 'main' into feat/AutoLockComponent

al-rosenthal committed:
fix: updated translation replace for auto lock settings

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

al-rosenthal committed:
fix: fixed snapshot test

Signed-off-by: al-rosenthal <alfred.k.rosenthal@gmail.com>

bryce-mcmath reviewed (approved):
Nice work!

jleach commented on packages/legacy/core/App/localization/fr/index.ts:
```diff
@@ -654,7 +654,14 @@ const translation = {
         "ScanMyQR": "Scanner mon code QR",
-        "Developer": "Options de développeur"
+        "Developer": "Options de développeur",
+        "AutoLockTime": "Auto lock time (fr)",
```
We tend to do this in all caps but not worth holding up the PR for.

jleach commented on packages/legacy/core/App/screens/PINEnter.tsx:
```diff
@@ -354,7 +354,9 @@ const PINEnter: React.FC<PINEnterProps> = ({ setAuthenticated, usage = PINEntryU
         <>
-          <Text style={style.helpText}>{t('PINEnter.LockedOut')}</Text>
+          <Text style={style.helpText}>
+            {t('PINEnter.LockedOut', { time: String(store.preferences.autoLockTime ?? 5) })}
```
I think 5 by itself is a magic number. In another PR circle back and define it somewhere like `const defautlAutolockInMinutes = 5`. Also, you have it here and in `    autoLockTime: 5, // default wallets lockout time to 5 minutes` so to change the default it needs to be updated in 2x places. Better to define it in constants and use it in both places.

jleach commented on packages/legacy/core/App/screens/AutoLock.tsx:
```diff
@@ -0,0 +1,135 @@
+    },
+  })
+
+  const handleTimeoutChange = (time: (typeof AutoLockTime)[keyof typeof AutoLockTime]) => {
```
No need for `useCallback`? If the linter is happy then so am I.
